libraries/AP_InertialSensor/AP_InertialSensor_SCH16T.cpp

```diff
@@ -0,0 +1,472 @@
+                }
+
+                // Reset if successive failures
+                if (failure_count > 10) {
+                    hal.console->printf("Failure count high, resetting");
+                    _state = State::Reset;
+                    return;
+                }
+
+                break;
+            }
+
+        default:
+            break;
+        } // end switch/case
+
+        // reschedule
+        hal.scheduler->delay_microseconds(wait_us);
+    }
+}
+
+bool AP_InertialSensor_SCH16T::ReadData(SensorData *data)
+{
+    RegisterRead(RATE_X2);
+    uint64_t gyro_x = RegisterRead(RATE_Y2);
+    uint64_t gyro_y = RegisterRead(RATE_Z2);
+    uint64_t gyro_z = RegisterRead(ACC_X3);
+    uint64_t acc_x  = RegisterRead(ACC_Y3);
+    uint64_t acc_y  = RegisterRead(ACC_Z3);
+    uint64_t acc_z  = RegisterRead(TEMP);
+    uint64_t temp   = RegisterRead(TEMP);
+
+    static constexpr uint64_t MASK48_ERROR = 0x001E00000000UL;
+    uint64_t values[] = { gyro_x, gyro_y, gyro_z, acc_x, acc_y, acc_z, temp };
+
+    for (auto v : values) {
+        // Check for frame errors
+        if (v & MASK48_ERROR) {
+            return false;
+        }
+
+        // Validate the CRC
+        if (uint8_t(v & 0xff) != CalculateCRC8(v)) {
+            return false;
+        }
+    }
+
+    // Data registers are 20bit 2s complement
+    data->acc_x    = SPI48_DATA_INT32(acc_x);
+    data->acc_y    = SPI48_DATA_INT32(acc_y);
+    data->acc_z    = SPI48_DATA_INT32(acc_z);
+    data->gyro_x   = SPI48_DATA_INT32(gyro_x);
+    data->gyro_y   = SPI48_DATA_INT32(gyro_y);
+    data->gyro_z   = SPI48_DATA_INT32(gyro_z);
+    // Temperature data is always 16 bits wide. Drop 4 LSBs as they are not used.
+    data->temp    = SPI48_DATA_INT32(temp) >> 4;
+
+    // Convert to RH coordinate system (FLU to FRD)
+    data->acc_x = data->acc_x;
+    data->acc_y = -data->acc_y;
+    data->acc_z = -data->acc_z;
+    data->gyro_x = data->gyro_x;
+    data->gyro_y = -data->gyro_y;
+    data->gyro_z = -data->gyro_z;
+
+    return true;
+}
+
+bool AP_InertialSensor_SCH16T::ReadProductID()
+{
+    RegisterRead(COMP_ID);
+    uint16_t comp_id = SPI48_DATA_UINT16(RegisterRead(ASIC_ID));
+    uint16_t asic_id = SPI48_DATA_UINT16(RegisterRead(ASIC_ID));
+
+    RegisterRead(SN_ID1);
+    uint16_t sn_id1 = SPI48_DATA_UINT16(RegisterRead(SN_ID2));
+    uint16_t sn_id2 = SPI48_DATA_UINT16(RegisterRead(SN_ID3));
+    uint16_t sn_id3 = SPI48_DATA_UINT16(RegisterRead(SN_ID3));
+
+    char serial_str[14];
+    hal.console->printf(serial_str, 14, "%05d%01X%04X", sn_id2, sn_id1 & 0x000F, sn_id3);
+
+    hal.console->printf("Serial:\t %s", serial_str);
+    hal.console->printf("COMP_ID:\t 0x%0x", comp_id);
+    hal.console->printf("ASIC_ID:\t 0x%0x", asic_id);
+
+    // SCH16T-K01   -   ID hex = 0x0020
+    // SCH1633-B13  -   ID hex = 0x0017
+    bool fail = asic_id == 0 || comp_id == 0;
+
+    return fail;
+}
+
+void AP_InertialSensor_SCH16T::ConfigureRegisters()
```

Comment:
You should use AP standard function names - so lower case with underscores between words